CDCgov/prime-reportstream
# Improvements and Fixes to the LookupTable API

### This PR does the following:
- allows lookuptable API calls to work with either server2server or Okta auth.
- fixes a bug that even lookuptable API _reads_ required PrimeAdmin access.   Now anyone authorized with ReportStream can do reads; however, PrimeAdmin is still required for Writes.
- the `"authentication-type: okta"` http header is no longer required. Yay!  Instead the code first tries to do okta auth.  If that fails to auth, it tries to do server2server auth.  If that fails, it gives up, and you are unauthorized.
- Added smoke tests around the lookup Table API.
- Added new helpful endpoint that defaults to latest active version of table if no version number is specified.  See the release notes for documentation.

### Test Steps:
Get an oktapreview token from a Staging **_PrimeAdmin_** user:
```
./prime login --env staging
```

Then run these smoke tests:

```
./prime test --run server2serverauth,otkaauth
```

You'll see at the end of each of those tests that it now runs some LookupTable auth tests, in addition to all the other auth tests.

Here are some example calls you can try by hand as well:
```
export TOK=$(cat ~/.prime/accesstoken.json | grep token | cut -d"," -f1 | cut -d":" -f2 | tr -d \")
curl -H "authorization: bearer $TOK" "http://localhost:7071/api/lookuptables/LIVD-SARS-CoV-2/1/info"
curl -H "authorization: bearer $TOK" "http://localhost:7071/api/lookuptables/LIVD-SARS-CoV-2/1/content" | wc
curl -H "authorization: bearer $TOK" "http://localhost:7071/api/lookuptables/list" | wc

# Should fail but not with an auth error
curl -X POST -H "authorization: bearer $TOK" "http://localhost:7071/api/lookuptables/foobar"
(actually I get  null pointer exception, which is a different problem!)

# SHould work:
# Be careful you don't auth an OLD livd table!!
curl -X PUT -H "authorization: bearer $TOK" "http://localhost:7071/api/lookuptables/LIVD-SARS-CoV-2/1/activate"

```
## Checklist

### Testing
- [N/A] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 

### Process
- [N] Are there licensing issues with any new dependencies introduced?
- [Y] Includes a summary of what a code reviewer should test/verify?
- [Y] Updated the release notes?
- [n/a] Database changes are submitted as a separate PR?
- [n/a] DevOps team has been notified if PR requires ops support?

## Linked Issues
- Fixes all of #6598
- Is related to parts of #5290 and #1946

## Specific Security-related subjects a reviewer should pay specific attention to

- Does this PR introduce new endpoints?
    - **_YES_**:   `api/lookuptables/TABLENAME/content`
- Does this PR include changes in authentication and/or authorization of existing endpoints?  **_YES_**

#### If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:
- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
 
**Answer:** 

The potential dangers would be be S(poofing) and T(ampering).  This PR does several auth related changes:   

1. First, it further spreads the use of the common combination okta/server2server auth to now include LookupTable endpoints.  Previously only the deliveries/submissions and data-upload ("waters") endpoints used that new combination auth pattern.   The pattern is:

```
    val claims = AuthenticatedClaims.authenticate(request)
                ?: return HttpUtilities.unauthorizedResponse(request, authenticationFailure)
    <gather more info, if needed>
    if (!doFunctionSpecificAuthorizationCheck(claims, <other info>)) {
               return HttpUtilities.unauthorizedResponse(request, authorizationFailure)
    }
    <execute  the function action>
```

The intention is that this standardized authn/authz style will spread to all our functions.   (See LookupTableFunctions.kt line 274 for an example of this style)

2. Second, it changes the way the authenticate call handles the two security schemes.  The old code used an http header parameter:

```
claims = if (authentication-type header parameter == "okta") 
     doOktaAuthentication()
else
    doServer2ServerAuthentication()
```

The new code eliminates the header parameter, with the cost that it always tries okta auth prior to attempting server2server auth:

```
claims = doOktaAuthentication() ?:  doServer2ServerAuthentication()
```
(see AuthenticatedClaims.kt line 215 for the actual code)
